iceoryx_utils/include/iceoryx_utils/cxx/expected.hpp

```diff
@@ -431,6 +433,10 @@ class expected<ValueType, ErrorType>
     template <typename... Targs>
     static expected create_error(Targs&&... args) noexcept;
 
+    /// @brief returns true if the expected is in a valid state either containing a value or an error
+    /// @return true if in valid state or false if in invalid state (e.g. after move)
+    bool is_initialized() const noexcept;
```

Comment:
I dislike this as now one could get the impression that we always need to check this before accessing the expected. I know other types also do it like this but there it is also bad. 
The convention is that it is the responsibility of the caller not to access any object that was moved from.
It can be silently invalidated (I also am not a fan of this, for performance and logic reasons) but it is important to not have a mandatory is_initialized check. 
I am fine with it in the public API, it can be used...but does not have to be.

Reply:
I agree with @MatthiasKillat .

And the more I think about the name `is_initialized` the more I dislike it. Since the user does not see right away when this method returns false they are going straight into the land of undefined behavior. When we perform a code review it has to shine so that we pay extra attention. What about the name `has_undefined_state` or `is_in_unsafe_undefined_state`?

Is there a way to hide this and make the expected still working without checking this?! I think this is possible by:
1. Adjusting the behavior of `operator bool()` and return true only if it is not in an unsafe undefined state and has a value
2. Add `bool has_value()` and `bool has_error()`. Then the user has to check the correct method when they would like to access the value/error - it is error prone! But if you would like to use it in the recommended safe way we have `and_then`/`or_else`.
3. `and_then` and `or_else` should never be called when the object is in an unsafe undefined state.

Reply:
At least two possbilities: the methods and_then, or_else etc. can either to do nothing when the object is not initialized (which leads to an internal check) or just have undefined behavior in this case (likely cashing). Since the second case can only occur for sources of move, it would be within how C++ works with moved objects, so no real surprises for the user.

has_value should be added regardless, to make this usage pattern more symmetric to optional.

Since we have ``operator bool``, is_initialized could be private (or removed) and the check added to ``operator bool``.

Note: I am completely fine with undefined behavior as long as it is in line with established C++ rules (i.e. cases where it occurs) and there is good reason for it (usually performance).